File: tests/java/org/apache/qpid/proton/InteropTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.qpid.proton;

import org.apache.qpid.proton.TestDecoder;
import org.apache.qpid.proton.amqp.Binary;
import org.apache.qpid.proton.amqp.DescribedType;
import org.apache.qpid.proton.amqp.Symbol;
import org.apache.qpid.proton.amqp.messaging.AmqpValue;
import org.apache.qpid.proton.message.Message;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertArrayEquals;
import org.junit.Test;
import java.lang.System;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.Vector;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;

public class InteropTest
{

    static private File findTestsInteropDir()
    {
        File f = new File(System.getProperty("user.dir"));
        while (f != null && !f.getName().equals("tests"))
            f = f.getParentFile();
        if (f != null && f.isDirectory())
            return new File(f, "interop");
        else
            throw new Error("Cannot find tests/interop directory");
    }

    static File testsInteropDir = findTestsInteropDir();

    byte[] getBytes(String name) throws IOException
    {
        File f = new File(testsInteropDir, name + ".amqp");
        byte[] data = new byte[(int) f.length()];
        FileInputStream fi = new FileInputStream(f);
        assertEquals(f.length(), fi.read(data));
        fi.close();
        return data;
    }

    Message decodeMessage(String name) throws IOException
    {
        byte[] data = getBytes(name);
        Message m = Proton.message();
        m.decode(data, 0, data.length);
        return m;
    }

    TestDecoder createDecoder(byte[] data)
    {
        TestDecoder td = new TestDecoder(data);
        return td;
    }

    @Test
    public void testMessage() throws IOException
    {
        Message m = decodeMessage("message");
        Binary b = (Binary) (((AmqpValue) m.getBody()).getValue());
        String s = createDecoder(b.getArray()).readString();
        assertEquals("hello", s);
    }

    @Test
    public void testPrimitives() throws IOException
    {
        TestDecoder d = createDecoder(getBytes("primitives"));
        assertEquals(true, d.readBoolean());
        assertEquals(false, d.readBoolean());
        assertEquals(d.readUnsignedByte().intValue(), 42);
        assertEquals(42, d.readUnsignedShort().intValue());
        assertEquals(-42, d.readShort().intValue());
        assertEquals(12345, d.readUnsignedInteger().intValue());
        assertEquals(-12345, d.readInteger().intValue());
        assertEquals(12345, d.readUnsignedLong().longValue());
        assertEquals(-12345, d.readLong().longValue());
        assertEquals(0.125, d.readFloat().floatValue(), 0e-10);
        assertEquals(0.125, d.readDouble().doubleValue(), 0e-10);
    }

    @Test
    public void testStrings() throws IOException
    {
        TestDecoder d = createDecoder(getBytes("strings"));
        assertEquals(new Binary("abc\0defg".getBytes("UTF-8")), d.readBinary());
        assertEquals("abcdefg", d.readString());
        assertEquals(Symbol.valueOf("abcdefg"), d.readSymbol());
        assertEquals(new Binary(new byte[0]), d.readBinary());
        assertEquals("", d.readString());
        assertEquals(Symbol.valueOf(""), d.readSymbol());
    }

    @Test
    public void testDescribed() throws IOException
    {
        TestDecoder d = createDecoder(getBytes("described"));
        DescribedType dt = (DescribedType) (d.readObject());
        assertEquals(Symbol.valueOf("foo-descriptor"), dt.getDescriptor());
        assertEquals("foo-value", dt.getDescribed());

        dt = (DescribedType) (d.readObject());
        assertEquals(12, dt.getDescriptor());
        assertEquals(13, dt.getDescribed());
    }

    @Test
    public void testDescribedArray() throws IOException
    {
        TestDecoder d = createDecoder(getBytes("described_array"));
        DescribedType a[] = (DescribedType[]) (d.readArray());
        for (int i = 0; i < 10; ++i)
        {
            assertEquals(Symbol.valueOf("int-array"), a[i].getDescriptor());
            assertEquals(i, a[i].getDescribed());
        }
    }

    @Test
    public void testArrays() throws IOException
    {
        TestDecoder d = createDecoder(getBytes("arrays"));

        // int array
        Vector<Integer> ints = new Vector<Integer>();
        for (int i = 0; i < 100; ++i)
            ints.add(new Integer(i));
        assertArrayEquals(ints.toArray(), d.readArray());

        // String array
        String strings[] =
        { "a", "b", "c" };
        assertArrayEquals(strings, d.readArray());

        // Empty array
        assertArrayEquals(new Integer[0], d.readArray());
    }

    @Test
    public void testLists() throws IOException
    {
        TestDecoder d = createDecoder(getBytes("lists"));
        List<Object> l = new ArrayList<Object>()
        {
            {
                add(new Integer(32));
                add("foo");
                add(new Boolean(true));
            }
        };
        assertEquals(l, d.readList());
        l.clear();
        assertEquals(l, d.readList());
    }

    @SuppressWarnings({ "rawtypes", "unchecked" })
    @Test
    public void testMaps() throws IOException
    {
        TestDecoder d = createDecoder(getBytes("maps"));
        Map map = new HashMap()
        {
            {
                put("one", 1);
                put("two", 2);
                put("three", 3);
            }
        };
        assertEquals(map, d.readMap());

        map = new HashMap()
        {
            {
                put(1, "one");
                put(2, "two");
                put(3, "three");
            }
        };
        assertEquals(map, d.readMap());

        map = new HashMap();
        assertEquals(map, d.readMap());
    }
}

File: proton-c/bindings/go/src/apache.org/proton/interop_test.go
/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/

// Test that conversion of Go type to/from AMQP is compatible with other
// bindings.
//
// FIXME aconway 2015-03-01: this should move to proton/tests/go when we integrate
// better with the proton build system.
//
package proton

import (
	"bytes"
	"io"
	"io/ioutil"
	"os"
	"reflect"
	"strings"
	"testing"
)

func getReader(name string) (r io.Reader) {
	r, err := os.Open("../../../../../../tests/interop/" + name + ".amqp")
	if err != nil {
		panic(errorf("Can't open %#v: %v", name, err))
	}
	return
}

func remaining(d *Decoder) string {
	remainder, _ := ioutil.ReadAll(io.MultiReader(d.Buffered(), d.reader))
	return string(remainder)
}

// assertDecode: want is the expected value, gotPtr is a pointer to a
// instance of the same type for Decode.
func assertDecode(d *Decoder, want interface{}, gotPtr interface{}) {
	err := d.Decode(gotPtr)
	if err != nil {
		panic(err)
	}
	got := reflect.ValueOf(gotPtr).Elem().Interface()
	if !reflect.DeepEqual(want, got) {
		panic(errorf("%T(%#v) != %T(%#v)", want, want, got, got))
	}
}

func TestUnmarshal(t *testing.T) {
	bytes, err := ioutil.ReadAll(getReader("strings"))
	if err != nil {
		t.Error(err)
	}
	for _, want := range []string{"abc\000defg", "abcdefg", "abcdefg", "", "", ""} {
		var got string
		n, err := Unmarshal(bytes, &got)
		if err != nil {
			t.Error(err)
		}
		if want != got {
			t.Errorf("%#v != %#v", want, got)
		}
		bytes = bytes[n:]
	}
}

func TestPrimitivesExact(t *testing.T) {
	d := NewDecoder(getReader("primitives"))
	// Decoding into exact types
	var b bool
	assertDecode(d, true, &b)
	assertDecode(d, false, &b)
	var u8 uint8
	assertDecode(d, uint8(42), &u8)
	var u16 uint16
	assertDecode(d, uint16(42), &u16)
	var i16 int16
	assertDecode(d, int16(-42), &i16)
	var u32 uint32
	assertDecode(d, uint32(12345), &u32)
	var i32 int32
	assertDecode(d, int32(-12345), &i32)
	var u64 uint64
	assertDecode(d, uint64(12345), &u64)
	var i64 int64
	assertDecode(d, int64(-12345), &i64)
	var f32 float32
	assertDecode(d, float32(0.125), &f32)
	var f64 float64
	assertDecode(d, float64(0.125), &f64)
}

func TestPrimitivesCompatible(t *testing.T) {
	d := NewDecoder(getReader("primitives"))
	// Decoding into compatible types
	var b bool
	var i int
	var u uint
	var f float64
	assertDecode(d, true, &b)
	assertDecode(d, false, &b)
	assertDecode(d, uint(42), &u)
	assertDecode(d, uint(42), &u)
	assertDecode(d, -42, &i)
	assertDecode(d, uint(12345), &u)
	assertDecode(d, -12345, &i)
	assertDecode(d, uint(12345), &u)
	assertDecode(d, -12345, &i)
	assertDecode(d, 0.125, &f)
	assertDecode(d, 0.125, &f)
}

func TestStrings(t *testing.T) {
	d := NewDecoder(getReader("strings"))
	// Test decoding as plain Go strings
	for _, want := range []string{"abc\000defg", "abcdefg", "abcdefg", "", "", ""} {
		var got string
		assertDecode(d, want, &got)
	}
	remains := remaining(d)
	if remains != "" {
		t.Errorf("leftover: %s", remains)
	}

	// Test decoding as specific string types
	d = NewDecoder(getReader("strings"))
	var bytes []byte
	var str, sym string
	assertDecode(d, []byte("abc\000defg"), &bytes)
	assertDecode(d, "abcdefg", &str)
	assertDecode(d, "abcdefg", &sym)
	assertDecode(d, make([]byte, 0), &bytes)
	assertDecode(d, "", &str)
	assertDecode(d, "", &sym)
	remains = remaining(d)
	if remains != "" {
		panic(errorf("leftover: %s", remains))
	}

	// Test some error handling
	d = NewDecoder(getReader("strings"))
	var s string
	err := d.Decode(s)
	if !strings.Contains(err.Error(), "not a pointer") {
		t.Error(err)
	}
	var i int
	err = d.Decode(&i)
	if !strings.Contains(err.Error(), "cannot unmarshal") {
		t.Error(err)
	}
}

func BenchmarkDecode(b *testing.B) {
	var buf bytes.Buffer
	for _, f := range []string{"strings", "primitives"} {
		_, err := buf.ReadFrom(getReader(f))
		if err != nil {
			panic(err)
		}
	}

	d := NewDecoder(bytes.NewReader(buf.Bytes()))

	decode := func(v interface{}) {
		err := d.Decode(v)
		if err != nil {
			panic(err)
		}
	}

	for i := 0; i < b.N; i++ {
		var by []byte
		// strings
		decode(&by)
		var s string
		decode(&s)
		decode(&s)
		decode(&by)
		decode(&s)
		decode(&s)
		// primitives
		var b bool
		decode(&b)
		decode(&b)
		var u8 uint8
		decode(&u8)
		var u16 uint16
		decode(&u16)
		var i16 int16
		decode(&i16)
		var u32 uint32
		decode(&u32)
		var i32 int32
		decode(&i32)
		var u64 uint64
		decode(&u64)
		var i64 int64
		decode(&i64)
		var f32 float32
		decode(&f32)
		var f64 float64
		decode(&f64)

		d = NewDecoder(bytes.NewReader(buf.Bytes()))
	}
}
